From 5de859c9c8f57588fb8e8991f9cc136a7238e66d Mon Sep 17 00:00:00 2001
From: Jonathan FROSDICK <jon.frosdick@st.com>
Date: Fri, 7 Nov 2008 13:48:15 +0000
Subject: [PATCH] rndis_host: fix transfer size negotiation

Author: Jean-Christophe Dubois <jcd@tribudubois.net>
Date:   Thu Mar 13 14:56:36 2008 -0800

    rndis_host: fix transfer size negotiation

    This patch should resolve a problem that's troubled support for
    some RNDIS peripherals.  It seems to have boiled down to using a
    variable to establish transfer size limits before it was assigned,
    which caused those devices to fallback to a default "jumbogram"
    mode we don't support.  Fix by assigning it earlier for RNDIS.

    Signed-off-by: Jean-Christophe Dubois <jcd@tribudubois.net>
    [ cleanups ]
    Signed-off-by: David Brownell <dbrownell@users.sourceforge.net>
    Signed-off-by: Jeff Garzik <jeff@garzik.org>

Signed-off-by: Stuart Menefy <stuart.menefy@st.com>
---
 drivers/net/usb/rndis_host.c |   11 ++++++++---
 1 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/net/usb/rndis_host.c b/drivers/net/usb/rndis_host.c
index cd991a0..bcac756 100644
--- a/drivers/net/usb/rndis_host.c
+++ b/drivers/net/usb/rndis_host.c
@@ -17,9 +17,6 @@
  * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
  */
 
-// #define	DEBUG			// error path messages, extra info
-// #define	VERBOSE			// more; success messages
-
 #include <linux/module.h>
 #include <linux/init.h>
 #include <linux/netdevice.h>
@@ -499,6 +496,14 @@ static int rndis_bind(struct usbnet *dev, struct usb_interface *intf)
 	net->hard_header_len += sizeof (struct rndis_data_hdr);
 	dev->hard_mtu = net->mtu + net->hard_header_len;
 
+        dev->maxpacket = usb_maxpacket(dev->udev, dev->out, 1);
+        if (dev->maxpacket == 0) {
+                if (netif_msg_probe(dev))
+                        dev_dbg(&intf->dev, "dev->maxpacket can't be 0\n");
+                retval = -EINVAL;
+                goto fail_and_release;
+        }
+
 	dev->rx_urb_size = dev->hard_mtu + (dev->maxpacket + 1);
 	dev->rx_urb_size &= ~(dev->maxpacket - 1);
 	u.init->max_transfer_size = cpu_to_le32(dev->rx_urb_size);
-- 
1.5.3.6

